drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+    driver constructor.
+    '''
+    def __init__(self, param_names, num_receivers):
+        self.num_receivers = num_receivers
+
+        load = lambda prop: setattr(self, prop, rospy.get_param('passive_sonar/' + prop))
+        [load(x) for x in param_names]
+
+        self.tf2_buf = tf2_ros.Buffer()
+
+        try:
+            self.ser = serial.Serial(port=self.port, baudrate=self.baud, timeout=self.read_timeout)
+            self.ser.flushInput()
+        except serial.SerialException, e:
+            rospy.err("Sonar serial connection error: " + str(e))
+            raise e
+
+    def input_request(self):
+        try:
+            self._request_signals()
+            self._receive_signals()
+            T = self.getReceiverPose(rospy.Time.now(), self.receiver_array_frame,
+                                     self.locating_frame)
+            self.translation, self.rotation = T
+            self.ready = True
+        except Exception as e:
+            rospy.logerr(str(e))
+            raise e
+ 
+    def _request_signals(self):
+        '''
+        Request a set of digital signals from a serial port
+
+        serial_port - open instance of serial.Serial
+        data_request_code - char or string to sent to receiver board to signal a data request
+        tx_start_code - char or string that the receiver board will send  us to signal the
+            start of a transmission
+        '''
+        self.ser.flushInput()
+        try:
+            readin = None
+
+            # Request raw signal tx until start bit is received
+            while readin == None or ord(readin) != ord(self.tx_start_code):
+                self.ser.write(self.tx_request_code)
+                readin = self.ser.read(1)
+                if len(readin) < len(self.tx_request_code):  # serial read timed out
+                    raise IOError('Timed out waiting for serial response.')
+
+        except serial.SerialException as e:
+            rospy.logerr(str(e))
+            return None
+
+    def _receive_signals(self):
+        '''
+        Receives a set of 1D signals from a serial port and packs them into a numpy array
+
+        serial_port - port of type serial.Serial from which to read
+        signal_size - number of sacalar elements in each 1D signal
+        scalar_size - size of each scalar in bytes
+        signal_bias - value to be subtracted from each read scalar before packing into array
+
+        returns: 2D numpy array with <num_signals> rows and <signal_size> columns
+        '''
+        self.signals = np.full((self.num_receivers, self.signal_size), self.signal_bias, dtype=float)
+
+        for channel in range(self.num_receivers):
+            for i in range(self.signal_size):
+                while self.ser.inWaiting() < self.scalar_size:
+                    pass
+                self.signals[channel, i] = float(self.ser.read(self.scalar_size)) - self.signal_bias
```

Comment:
Does the board send 32 bit floats or 64 bit floats, I think python does everything in 64? It seems like this could cause confusion or problems at some point. It might be good to use `np.float64(...)` to make it clear what kind of data it's actually reading.

Also wouldn't it be good to do some CRC checking?